Logs: Merge Log Volumes by data source name

ifrost opened the pull request:
**What is this feature?**

As requested by @grafana/observability-logs, this PR introduces merging logs volume by data source name to bring back the behavior that was changed by #63036.

Log volumes coming from the data source with the same name are aggregated into a single view:

| before | after |
|---|---|
|  <img width="1528" alt="Screenshot 2023-03-27 at 22 00 40" src="https://user-images.githubusercontent.com/745532/228053843-c39bccc8-5b94-46a3-bd76-8e1ea711d679.png"> | <img width="1532" alt="Screenshot 2023-03-27 at 22 03 30" src="https://user-images.githubusercontent.com/745532/228053881-f73e9f76-895d-49bb-9ed7-9211dc1f489d.png"> |
|  <img width="1530" alt="Screenshot 2023-03-27 at 22 01 08" src="https://user-images.githubusercontent.com/745532/228053958-1788dd6b-a33e-432d-a661-31cc605da757.png"> |<img width="1536" alt="Screenshot 2023-03-27 at 22 03 16" src="https://user-images.githubusercontent.com/745532/228053995-eb98cc8b-00a5-4c92-ba65-e5e0662ffdcb.png"> |

Fixes #61546

**Special notes for your reviewer:**

I wanted to bring back the aggregation logic that was removed in #63036 but I realized it was wrong. It was aggregating data points by point index, not by timestamp :O  (it was based on wrong assumption that all data frames have the same length but that's not always the case, the length of data frames may be different because each data frame starts with the first existing data point, not with the beginning of the selected range)

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.
- [ ] There are no known compatibility issues with older supported versions of Grafana, or plugins.
- [ ] It passes the [Hosted Grafana feature readiness review](https://docs.google.com/document/d/1QL9Ly8KnXzpb6ISbg49pTODRO5mhA5tkkfIZVX6pqQU/edit) for observability, scalability, performance, and security.

ifrost committed:
Merge log volume by data source name

ifrost committed:
Fix creating response for multiple fallback volumes

ifrost committed:
Fix unit tests

ifrost committed:
Hide title if there's only one log volume visible

ifrost committed:
Make hide title optional

ifrost commented on public/app/features/explore/utils/supplementaryQueries.ts:
```diff
@@ -84,9 +85,11 @@ const createFallbackLogVolumeProvider = (
         const bucketSize = exploreData.logsResult.bucketSize;
         const targetRefIds = queryTargets.map((query) => query.refId);
         const rowsByRefId = groupBy(exploreData.logsResult.rows, 'dataFrame.refId');
+        let allSeries: DataFrame[] = [];
```
The aggregation for the fallback was emitting only the last value instead of aggregating them. I've added a new test `Creates single fallback result` that covers for it (it'd fail without these changes).

ivanahuckova reviewed (commented):
LGTM! I've left 1 small comment, but  mostly I am just curious. Also I discovered a bug - `onHiddenSeriesChanged` doesn't work because the levels now come in `{level="debug"}` form and not as strings. 

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/30407135/228226731-04785ed4-50bc-481b-b02e-fb1834170bba.png">

ifrost committed:
Remove redundant parentheses

ifrost commented on public/app/features/explore/LogsVolumePanel.tsx:
```diff
@@ -43,7 +44,7 @@ export function LogsVolumePanel(props: Props) {
 
   const logsVolumeInfo = getLogsVolumeDataSourceInfo(logsVolumeData?.data);
-  let extraInfo = logsVolumeInfo ? `${logsVolumeInfo.refId} (${logsVolumeInfo.name})` : '';
+  let extraInfo = logsVolumeInfo ? `(${logsVolumeInfo.name})` : '';
```
No reason 🤷🏻 Updated in ab47fdea19c5f0f84c8b678fa7933246483dd9b8

ifrost committed:
Do not use frame.name, so the visualization can pick displayNameFromDS from the field config

ifrost commented:
> Also I discovered a bug - `onHiddenSeriesChanged` doesn't work because the levels now come in `{level="debug"}` form and not as strings.

Good spot! Fixed in 9f0887ea6fdf5a9dbc042790982d35501e9311e8.  `frame.name` was set overriding the display name from field config; I've removed it.

ivanahuckova reviewed (commented):
I found 1 more issue with legend name in test data source - they all have name `Value` and it is also missing `This datasource does not support full-range histograms. The graph below is based on the logs seen in the response.`

This branch:
<img width="1475" alt="image" src="https://user-images.githubusercontent.com/30407135/228267312-18f318c3-bf3a-48bd-a96a-9731de4fb3bc.png">

Main branch:

<img width="1838" alt="image" src="https://user-images.githubusercontent.com/30407135/228267628-5422fc9a-63c2-4e79-af89-01fe504d3089.png">

ifrost committed:
Merge branch 'main' into ifrost/merge-log-volume

Elfo404 commented:
@ivanahuckova that is weird, i'm on this branch and i can see the labels correctly named in every scenario i tested (both mixed and non, with and without multiple datasaources / queries, including using testdata) 🤔 

but i can confirm the missing support for full range volume is not showing up

ifrost committed:
Simplify setting aggregated data frame meta data

ifrost commented on public/app/features/logs/utils.ts:
```diff
@@ -150,3 +162,67 @@ export function logRowsToReadableJson(logs: LogRowModel[]) {
+    configs[level] = {
+      meta: { ...(dataFrame.meta || {}) },
+      valueFieldConfig: { ...valueField.config },
+      timeFieldConfig: { ...timeField.config },
```
Updated in 1687169b3442db98c47682a8a2cbce54ab79fb00

ifrost committed:
Update public/app/features/logs/utils.ts

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

ifrost committed:
Fix legend toggling

ifrost committed:
Ensure limited graph info is shown

ifrost commented:
Thanks @ivanahuckova!

> I found 1 more issue with legend name in test data source - they all have name Value and

That's actually interestin. I went with this back and forth like a lunatic. Turns out it's a bit confusing how the legend and toggling are built out of the data frame -> it checks for `frame.name` first then `frame.fields[].config.displayNameFromDS`.

I've added code to set aggregated `frame.name` to level in 9fa38ff985e464c99e653f53ec77d11fe94ac11c to ensure it's consistent.

Just for context how it behaved in the past:

* 9.4.x: 
  * Loki's full-range log volume returns frames with names like `{level="debug"}` but it's aggregated by creating `MutableDataFrame` which omits the `name` so the legend falls back to `frame.fields[].config.displayNameFromDS` -everything is working fine 👍 
* `main`: (no merging, grouping by query)
  * Loki's full-range log volume is not merged, meaning `frame.name` is preserved, incorrect name (e.g. `{level="debug"}` instead of `debug`) is passed and toggling doesn't work -> **Yes, It's actually broken in main** 😢 
* `this branch`
  * Loki's full-range log volume is merged again so it's back to normal, but since limited log volume goes through the same logic now we need to re-add `frame.name` to newly created aggregated frame so it's not lost. Everything should work fine again ☕ 
 
It should now work nicely with limited and non-limited log volumes with the change added in 9fa38ff985e464c99e653f53ec77d11fe94ac11c.

> it is also missing This datasource does not support full-range histograms. The graph below is based on the logs seen in the response.

Oh yes, I was trying to be smart and hide the title when there's only one log volume displayed but this also removed the info about the limited graph which is added to the title 🤦🏻 Fixed in 59af43ce434254fcb26cb93b9d403a26cf48e952

ivanahuckova reviewed (commented):
Thank you for going into details on what is happening. Very insightful! 🙂

I have found 1 last small issue related to queries with single data source that does not support full range logs volume - now you get `. ` before the message about logs volume. I personally wouldn't mind if the name of data source would be always there, but up to you on how do you want to resolve it. 🙂

<img width="1439" alt="image" src="https://user-images.githubusercontent.com/30407135/228480568-f47b0649-449d-4efc-b562-744408ebe710.png">

ifrost committed:
Always show the data source name

ifrost commented:
> I have found 1 last small issue related to queries with single data source that does not support full range logs volume - now you get `. ` before the message about logs volume. I personally wouldn't mind if the name of data source would be always there, but up to you on how do you want to resolve it. 🙂

I actually like always showing the name. With mixed data source the user can for example run a log query and metric query from different data sources. That would produce a single log volume histogram, and still useful to indicate which data source the data is coming from. Changed in 678b34537c6e433c022b194a79f0fef07e43cca3
